File: src/services/stockApi.ts
/**
 * Stock API Service
 * 
 * Browser-compatible stock market data service
 * Uses multiple data sources with fallbacks
 */

export class StockApiService {
  private static instance: StockApiService;
  private cache: Map<string, { data: any; timestamp: number }> = new Map();
  private readonly CACHE_TTL = 60000; // 1 minute cache

  private constructor() {}

  static getInstance(): StockApiService {
    if (!StockApiService.instance) {
      StockApiService.instance = new StockApiService();
    }
    return StockApiService.instance;
  }

  private getCached(key: string): any | null {
    const cached = this.cache.get(key);
    if (cached && Date.now() - cached.timestamp < this.CACHE_TTL) {
      return cached.data;
    }
    return null;
  }

  private setCache(key: string, data: any): void {
    this.cache.set(key, { data, timestamp: Date.now() });
  }

  /**
   * Search for stocks by query
   * Uses Yahoo Finance search endpoint
   */
  async searchStocks(query: string): Promise<any[]> {
    try {
      const cacheKey = `search_${query}`;
      const cached = this.getCached(cacheKey);
      if (cached) return cached;

      // Use Yahoo Finance search via public API
      const url = `https://query2.finance.yahoo.com/v1/finance/search?q=${encodeURIComponent(query)}&quotesCount=10&newsCount=0`;
      
      const response = await fetch(url);
      if (!response.ok) throw new Error('Search failed');

      const data = await response.json();
      const results = (data.quotes || [])
        .filter((q: any) => q.quoteType === 'EQUITY' && q.symbol)
        .slice(0, 10)
        .map((q: any) => ({
          symbol: q.symbol,
          shortName: q.shortname || q.longname || q.symbol,
          longName: q.longname || q.shortname || q.symbol,
          exchange: q.exchange || 'UNKNOWN',
        }));

      this.setCache(cacheKey, results);
      return results;
    } catch (error) {
      console.error('Error searching stocks:', error);
      return [];
    }
  }

  /**
   * Get current quote for a stock
   * Uses Yahoo Finance quote endpoint
   */
  async getQuote(symbol: string): Promise<any | null> {
    try {
      const cacheKey = `quote_${symbol}`;
      const cached = this.getCached(cacheKey);
      if (cached) return cached;

      // Use Yahoo Finance quote API
      const url = `https://query1.finance.yahoo.com/v8/finance/chart/${symbol}?interval=1d&range=1d`;
      
      const response = await fetch(url);
      if (!response.ok) throw new Error('Quote fetch failed');

      const data = await response.json();
      const result = data.chart?.result?.[0];
      
      if (!result || !result.meta) {
        return null;
      }

      const meta = result.meta;
      const quote = {
        symbol: meta.symbol,
        shortName: meta.shortName || symbol,
        longName: meta.longName || meta.shortName || symbol,
        regularMarketPrice: meta.regularMarketPrice || 0,
        regularMarketChange: meta.regularMarketPrice - (meta.previousClose || meta.regularMarketPrice) || 0,
        regularMarketChangePercent: meta.regularMarketPrice && meta.previousClose
          ? ((meta.regularMarketPrice - meta.previousClose) / meta.previousClose) * 100
          : 0,
        regularMarketPreviousClose: meta.previousClose || 0,
        regularMarketOpen: meta.regularMarketOpen || 0,
        regularMarketDayHigh: meta.regularMarketDayHigh || 0,
        regularMarketDayLow: meta.regularMarketDayLow || 0,
        regularMarketVolume: meta.regularMarketVolume || 0,
        marketCap: meta.marketCap,
        fiftyTwoWeekHigh: meta.fiftyTwoWeekHigh || 0,
        fiftyTwoWeekLow: meta.fiftyTwoWeekLow || 0,
        currency: meta.currency || 'USD',
        exchange: meta.exchangeName || meta.exchange || 'UNKNOWN',
      };

      this.setCache(cacheKey, quote);
      return quote;
    } catch (error) {
      console.error(`Error fetching quote for ${symbol}:`, error);
      return null;
    }
  }

  /**
   * Get historical data for a stock
   * Uses Yahoo Finance historical data endpoint
   */
  async getHistoricalData(
    symbol: string,
    period: '1d' | '5d' | '1mo' | '3mo' | '6mo' | '1y' | '2y' | '5y' | 'max' = '1mo'
  ): Promise<any[]> {
    try {
      const cacheKey = `historical_${symbol}_${period}`;
      const cached = this.getCached(cacheKey);
      if (cached) return cached;

      const range = period === 'max' ? '10y' : period;
      const interval = this.getInterval(period);
      
      // Use Yahoo Finance chart API
      const url = `https://query1.finance.yahoo.com/v8/finance/chart/${symbol}?interval=${interval}&range=${range}`;
      
      const response = await fetch(url);
      if (!response.ok) throw new Error('Historical data fetch failed');

      const data = await response.json();
      const result = data.chart?.result?.[0];
      
      if (!result || !result.timestamp || !result.indicators) {
        return [];
      }

      const timestamps = result.timestamp || [];
      const quotes = result.indicators?.quote?.[0] || {};
      const opens = quotes.open || [];
      const highs = quotes.high || [];
      const lows = quotes.low || [];
      const closes = quotes.close || [];
      const volumes = quotes.volume || [];

      const formattedData = timestamps
        .map((timestamp: number, index: number) => {
          const date = new Date(timestamp * 1000);
          return {
            date: date.toISOString().split('T')[0],
            open: opens[index] || 0,
            high: highs[index] || 0,
            low: lows[index] || 0,
            close: closes[index] || 0,
            volume: volumes[index] || 0,
            adjustedClose: closes[index] || 0,
          };
        })
        .filter((item: any) => item.close > 0)
        .reverse(); // Most recent first

      this.setCache(cacheKey, formattedData);
      return formattedData;
    } catch (error) {
      console.error(`Error fetching historical data for ${symbol}:`, error);
      return [];
    }
  }

  /**
   * Get multiple quotes at once
   */
  async getMultipleQuotes(symbols: string[]): Promise<any[]> {
    try {
      // Fetch quotes in parallel with limited concurrency
      const batchSize = 5;
      const quotes: any[] = [];
      
      for (let i = 0; i < symbols.length; i += batchSize) {
        const batch = symbols.slice(i, i + batchSize);
        const batchQuotes = await Promise.all(
          batch.map(symbol => this.getQuote(symbol))
        );
        quotes.push(...batchQuotes.filter(quote => quote !== null));
        
        // Small delay between batches to avoid rate limiting
        if (i + batchSize < symbols.length) {
          await new Promise(resolve => setTimeout(resolve, 200));
        }
      }
      
      return quotes;
    } catch (error) {
      console.error('Error fetching multiple quotes:', error);
      return [];
    }
  }

  private getInterval(period: string): '1d' | '1h' | '5m' | '15m' | '30m' | '1m' {
    if (period === '1d') return '5m';
    if (period === '5d') return '15m';
    if (period === '1mo') return '1d';
    return '1d';
  }
}

export const stockApi = StockApiService.getInstance();

File: backend/app/services/stock_api.py
"""
Stock API Service - Integrates with Alpha Vantage and Yahoo Finance
"""

from datetime import datetime
from typing import Dict, List, Optional

import httpx

from app.core.config import settings


class StockAPIService:
    """Service for fetching stock data from various APIs"""

    def __init__(self):
        self.alpha_vantage_key = settings.ALPHA_VANTAGE_API_KEY
        self.yahoo_finance_enabled = settings.YAHOO_FINANCE_ENABLED
        self.base_url_alpha = "https://www.alphavantage.co/query"
        self.base_url_yahoo = (
            "https://query1.finance.yahoo.com/v8/finance/chart"
        )
        # Lazily-created AsyncClient reused across requests to avoid
        # repeatedly creating/closing connections which harms performance.
        self._client = None

    async def _get_client(self):
        """Return a shared AsyncClient instance, creating it if needed."""
        if self._client is None:
            self._client = httpx.AsyncClient(timeout=10.0)
        return self._client

    async def get_quote(self, symbol: str) -> Optional[Dict]:
        """Get current stock quote"""
        try:
            # Try Alpha Vantage first if API key is available
            if self.alpha_vantage_key:
                quote = await self._get_alpha_vantage_quote(symbol)
                if quote:
                    return quote

            # Fallback to Yahoo Finance
            if self.yahoo_finance_enabled:
                return await self._get_yahoo_quote(symbol)

            return None
        except Exception as e:
            print(f"Error fetching quote for {symbol}: {str(e)}")
            return None

    async def _get_alpha_vantage_quote(self, symbol: str) -> Optional[Dict]:
        """Get quote from Alpha Vantage"""
        try:
            client = await self._get_client()
            params = {
                "function": "GLOBAL_QUOTE",
                "symbol": symbol,
                "apikey": self.alpha_vantage_key,
            }
            response = await client.get(self.base_url_alpha, params=params)
            data = response.json()

                if "Global Quote" in data:
                    quote = data["Global Quote"]
                    return {
                        "symbol": quote.get("01. symbol"),
                        "short_name": quote.get("01. symbol"),
                        "price": float(quote.get("05. price", 0)),
                        "change": float(quote.get("09. change", 0)),
                        "change_percent": float(
                            quote.get(
                                "10. change percent", "0%"
                            ).replace("%", "")
                        ),
                        "volume": int(quote.get("06. volume", 0)),
                        "currency": "USD",
                        "exchange": "NASDAQ",
                    }
        except Exception as e:
            print(f"Alpha Vantage error: {str(e)}")

        return None

    async def _get_yahoo_quote(self, symbol: str) -> Optional[Dict]:
        """Get quote from Yahoo Finance (no API key required)"""
        try:
            client = await self._get_client()
            url = f"{self.base_url_yahoo}/{symbol}"
            params = {"interval": "1d", "range": "1d"}
            response = await client.get(url, params=params)
            data = response.json()

                if "chart" in data and "result" in data["chart"]:
                    result = data["chart"]["result"][0]
                    meta = result.get("meta", {})

                    current_price = meta.get("regularMarketPrice", 0)
                    previous_close = meta.get("previousClose", current_price)
                    change = current_price - previous_close
                    change_percent = (
                        (change / previous_close * 100)
                        if previous_close
                        else 0
                    )

                    return {
                        "symbol": meta.get("symbol", symbol),
                        "short_name": meta.get("shortName", symbol),
                        "long_name": meta.get("longName"),
                        "price": current_price,
                        "change": change,
                        "change_percent": change_percent,
                        "volume": meta.get("regularMarketVolume", 0),
                        "market_cap": meta.get("marketCap"),
                        "currency": meta.get("currency", "USD"),
                        "exchange": meta.get("exchange", "NASDAQ"),
                    }
        except Exception as e:
            print(f"Yahoo Finance error: {str(e)}")

        return None

    async def get_historical_data(
        self, symbol: str, period: str = "1mo"
    ) -> List[Dict]:
        """Get historical stock data"""
        try:
            if self.yahoo_finance_enabled:
                return await self._get_yahoo_historical(symbol, period)
            return []
        except Exception as e:
            print(f"Error fetching historical data: {str(e)}")
            return []

    async def _get_yahoo_historical(
        self, symbol: str, period: str
    ) -> List[Dict]:
        """Get historical data from Yahoo Finance"""
        try:
            client = await self._get_client()
            url = f"{self.base_url_yahoo}/{symbol}"
            params = {"interval": "1d", "range": period}
            response = await client.get(url, params=params)
            data = response.json()

                if "chart" in data and "result" in data["chart"]:
                    result = data["chart"]["result"][0]
                    timestamps = result.get("timestamp", [])
                    indicators = result.get("indicators", {})
                    quote = indicators.get("quote", [{}])[0]

                    opens = quote.get("open", [])
                    highs = quote.get("high", [])
                    lows = quote.get("low", [])
                    closes = quote.get("close", [])
                    volumes = quote.get("volume", [])

                    historical_data = []
                    for i, ts in enumerate(timestamps):
                        if i < len(closes) and closes[i]:
                            historical_data.append(
                                {
                                    "date": datetime.fromtimestamp(
                                        ts
                                    ).isoformat(),
                                    "open":
                                        opens[i]
                                        if i < len(opens)
                                        else closes[i],
                                    "high":
                                        highs[i]
                                        if i < len(highs)
                                        else closes[i],
                                    "low":
                                        lows[i]
                                        if i < len(lows)
                                        else closes[i],
                                    "close": closes[i],
                                    "volume": (
                                        int(volumes[i])
                                        if i < len(volumes)
                                        else 0
                                    ),
                                }
                            )

                    return historical_data
        except Exception as e:
            print(f"Yahoo Finance historical error: {str(e)}")

        return []

    async def search_stocks(self, query: str) -> List[Dict]:
        """Search for stocks by symbol or company name"""
        try:
            # For now, return mock data or use a search API
            # In production, integrate with a proper stock search API
            client = await self._get_client()
            # Using Yahoo Finance search
            url = "https://query1.finance.yahoo.com/v1/finance/search"
            params = {"q": query, "quotesCount": 10}
            response = await client.get(url, params=params)
            data = response.json()

                results = []
                if "quotes" in data:
                    for quote in data["quotes"]:
                        results.append(
                            {
                                "symbol": quote.get("symbol"),
                                "short_name": quote.get("shortname"),
                                "long_name": quote.get("longname"),
                                "exchange": quote.get("exchange", "NASDAQ"),
                            }
                        )

                return results
        except Exception as e:
            print(f"Error searching stocks: {str(e)}")
            return []

    async def get_multiple_quotes(self, symbols: List[str]) -> List[Dict]:
        """Get quotes for multiple symbols"""
        quotes = []
        for symbol in symbols:
            quote = await self.get_quote(symbol)
            if quote:
                quotes.append(quote)
        return quotes


# Singleton instance
stock_api_service = StockAPIService()
